projects/dsp-ui/README.md

```diff
@@ -1,26 +1,288 @@
+
+```typescript
+export const environment = {
+  name: 'dev',      // <-- add here the name 'dev' or 'prod' etc.
+  production: false
+};
+```
+
+The config files have to been integrated in `angular.json` in all "assets"-sections:
+
+```json
+"assets": [
+    "src/favicon.ico",
+    "src/assets",
+    "src/config"    <-- add this line and do not forget the comma in previous line
+]
+```
+
+Then we can make a service that will fetch the configs:
+
+```shell
+ng generate service app-init
+```
+
+```typescript
+import { Injectable } from '@angular/core';
+import { KnoraApiConfig, KnoraApiConnection } from '@knora/api';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AppInitService {
+
+  static dspApiConnection: KnoraApiConnection;
+
+  static dspApiConfig: KnoraApiConfig;
+
+  constructor() { }
+
+  Init() {
+
+    return new Promise<void>((resolve) => {
+
+      // get api config information from temp storage
+      const dspApiConfig: KnoraApiConfig = window['tempConfigStorage'] as KnoraApiConfig;
+
+      // init dsp-api configuration
+      AppInitService.dspApiConfig = new KnoraApiConfig(
+        dspApiConfig.apiProtocol,
+        dspApiConfig.apiHost,
+        dspApiConfig.apiPort,
+        dspApiConfig.apiPath,
+        dspApiConfig.jsonWebToken,
+        dspApiConfig.logErrors
+      );
+
+      // set knora-api connection configuration
+      AppInitService.dspApiConnection = new KnoraApiConnection(AppInitService.dspApiConfig);
+
+      resolve();
+    });
+  }
+}
+```
+
+Once we have this service, we have to edit the `app.module.ts` with a function to load the config:
+
+```typescript
+export function initializeApp(appInitService: AppInitService) {
+  return (): Promise<any> => {
+    return appInitService.Init();
+  };
+}
+```
+
+Provide it in the main module and include the desired DSP-UI modules in the imports:
+
+```typescript
+@NgModule({
+  declarations: [
+    AppComponent,
+    ProjectsComponent
+  ],
+  imports: [
+    BrowserModule,
+    AppRoutingModule,
+    DspCoreModule,      // <-- add the dsp-ui modules here
+    DspViewerModule
+  ],
+  providers: [
+      // add the following code block
+    AppInitService,
+    {
+      provide: APP_INITIALIZER,
+      useFactory: initializeApp,
+      deps: [AppInitService],
+      multi: true
+    },
+    {
+      provide: DspApiConfigToken,
+      useFactory: () => AppInitService.dspApiConfig
+    },
+    {
+      provide: DspApiConnectionToken,
+      useFactory: () => AppInitService.dspApiConnection
+    }
+  ],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+```
+
+Do not forget to make the imports for `APP_INITIALIZER` from `@angular/core` and the desired DSP-UI modules from `@dasch-swiss/dsp-ui`.
+
+Finally, the `main.ts` file must be modified:
```

Comment:
could you explain why it has to be modified?

Reply:
Something like:

"Finally, the `main.ts` file must be modified to load the environment specific config file and to test, that the config is correct:"

Reply:
Wasn't this a timing problem (so that the config is loaded when the app starts)? And the only way to do this was changing the `main.ts`. @subotic Is this correct?

I think the config file's contents are read and stored in a global variable `window['tempConfigStorage']` and then read back from there in the `app.init.service.ts`.

Reply:
Yes, we have to load the configuration when loading the app. This is done with several things:
- APP_INITIALIZER in app.modules
- app-init.service
- and main.ts to store in tempConfigStorage

It's also possible to make an http request to the config file, as some tutorials (e.g.: https://nbe.io/load-a-configuration-at-runtime-with-angular/) describe, but we decided to use the window.tempConfigStorage. This solution is more flexible and we can still set several configurations for various environments.

Reply:
ok, so just explain this briefly. Then it's clear why there is a need to change `main.ts` in the app.